# ENCODE

## ENCODE Blacklists

References
- Paper: Amemiya HM, Kundaje A, Boyle AP. The ENCODE Blacklist: Identification of Problematic Regions of the Genome. *Sci Rep*. 2019;9(1):9354. Published 2019 Jun 27. doi:[10.1038/s41598-019-45839-z](https://doi.org/10.1038/s41598-019-45839-z)
- Anshul Kundaje's webpage: https://sites.google.com/site/anshulkundaje/projects/blacklists
- ENCODE Annotation File Set: https://www.encodeproject.org/annotations/ENCSR636HFF/
- Boyle lab GitHub repo: https://github.com/Boyle-Lab/Blacklist

| Genome | Blacklist version | Links |
| ------ | ----------------- | ----- |
| hg19   | v1                | [GitHub](https://github.com/Boyle-Lab/Blacklist/raw/master/lists/Blacklist_v1/hg19-blacklist.bed.gz), [ENCODE](https://www.encodeproject.org/files/ENCFF001TDO/@@download/ENCFF001TDO.bed.gz)* |
| hg19   | v2                | [GitHub](https://github.com/Boyle-Lab/Blacklist/raw/master/lists/hg19-blacklist.v2.bed.gz) |
| GRCh38 | v1                | [Kundaje](http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/hg38-human/hg38.blacklist.bed.gz), [GitHub](https://github.com/Boyle-Lab/Blacklist/raw/master/lists/Blacklist_v1/hg38-blacklist.bed.gz) |
| GRCh38 | v2                | [GitHub](https://github.com/Boyle-Lab/Blacklist/raw/master/lists/hg38-blacklist.v2.bed.gz) |
| GRCh38 | v3                | [ENCODE](https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz) |
| mm10   | v1                | [ENCODE](https://www.encodeproject.org/files/ENCFF547MET/@@download/ENCFF547MET.bed.gz), [Kundaje](http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz), [GitHub](https://github.com/Boyle-Lab/Blacklist/raw/master/lists/Blacklist_v1/mm10-blacklist.bed.gz) |
| mm10   | v2                | [GitHub](https://github.com/Boyle-Lab/Blacklist/raw/master/lists/mm10-blacklist.v2.bed.gz)

\* Confusingly, [Anshul Kundaje's webpage](https://sites.google.com/site/anshulkundaje/projects/blacklists) lists the hg19 annotation file [ENCFF001TDO](https://www.encodeproject.org/files/ENCFF001TDO/@@download/ENCFF001TDO.bed.gz) as both Version 1 and Version 3. The file is identical to Version 1 of the hg19 blacklist on the Boyle Lab GitHub.

# UCSC

Download site: https://hgdownload.soe.ucsc.edu/downloads.html

Genome-specific data: `https://hgdownload.soe.ucsc.edu/goldenPath/<genome>/`
- Example: Human GRCh38 = https://hgdownload.soe.ucsc.edu/goldenPath/hg38/
- Structure
  - chromosomes/: a FASTA file for each chromosome
  - database/: annotations
    - RepeatMasker tracks
      - rmsk.txt.gz: source unclear (not described in the GitHub repo)
      - rmskAlign* and rmskOut*: very similar, except that Align corresponds to the ".align" file generated by RepeatMasker that shows the alignments between the repeat and query sequence, which is missing from the ".out" file. rmskOut* appears to be the main annotation file to use.
      - rmsk\*Baseline vs. rmsk\*Current: based on the GitHub repo, appear to correspond to older vs. newer annotations
      - rmskJoined*: a "RepeatMasker Visualization track" (unclear)
    - ... many others ...
  - bigZips/: genome, selected annotation files and updates
    - RepeatMasker-masked genome FASTA files
    - hg38.chrom.sizes
    - (Updated regularly) RefSeq mRNA multi-FASTA files
    - (Updated regularly) upstream1000/2000/5000: sequences 1000/2000/5000 bases upstream of annotated TSSs of RefSeq genes with annotated 5' UTRs.

Provenance of files: https://github.com/ucscGenomeBrowser/kent/tree/master/src/hg/makeDb/doc
- Example: code for main Human GRCh38 annotations = https://github.com/ucscGenomeBrowser/kent/blob/master/src/hg/makeDb/doc/hg38/hg38.txt

# NCI Genomic Data Commons

Abbreviations
- NCI: National Cancer Institute
- GDC: Genomic Data Commons

Documentation: https://docs.gdc.cancer.gov/

## Searching and filtering data

Available fields: https://docs.gdc.cancer.gov/API/Users_Guide/Appendix_A_Available_Fields/
- Definitions of each field can be found by searching the last component of the field name (i.e., after the last period in the field name) in the  in the [GDC Data Dictionary](https://docs.gdc.cancer.gov/Data_Dictionary/gdcmvs/).

## TCGA Barcode

Abbreviations
- TSS: Tissue Source Site
- BCR: Biospecimen Core Resource

Format: `[project]-[TSS]-[participant]-[sample][vial]-[portion][analyte]-[plate]-[center]`

When retrieving data via the [GDC API](https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/), the TCGA Barcode can be found (if applicable) under the `cases.samples.submitter_id`.

Reference
1. GDC Documentation: https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/
2. TCGA Code Tables: https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables
3. Wikipedia: https://en.wikipedia.org/wiki/The_Cancer_Genome_Atlas
