# Directory Setup

igv-launcher_portable.bat: adapted from igv-launcher.bat with the following key changes
  1. Ignores `%USERPROFILE%/.igv/java_arguments`
  2. Uses/stores preferences, log file, and genomes in `data/` directory

data/:
- genomes
  - Stores genome specification files. See [Downloading Hosted Genomes for Offline Use](https://github.com/igvteam/igv/wiki/Downloading-Hosted-Genomes-for-Offline-Use) and [JSON Genome Format](https://github.com/igvteam/igv/wiki/JSON-Genome-Format).
    - Note that binary \*.genome files (now considered deprecated) are just zip archives [(IGV documentation)](https://software.broadinstitute.org/software/igv/configuring_genome_server). They are not necessarily complete genomes for local (offline) access and can still request online data.
      - The \*.genome files available via the IGV/Broad Institute Amazon S3 buckets do *not* appear to contain the actual genome sequences; they appear to only contain some gene annotations.
  - The hg38/mm10/chm13v2.0.json files are from the `igv.org.genomes` Amazon S3 bucket (`<genome>/<genome>.json`).
    - The RefSeq tracks are [genePred](https://genome.ucsc.edu/FAQ/FAQformat.html#format9) format files generated by UCSC. These are not indexed, so they are completely loaded into memory. That is okay, since the files are only tens of MB uncompressed.
- igv0.log: log file
- prefs.properties: preferences file

# IGV Notes

Default behaviors of IGV version 2.13.2
- If no genome is available in its data directory (see `--igvDirectory` argument; defaults to `"%USERPROFILE%/igv/"`), it will try to load hg19 from https://s3.amazonaws.com/igv.org.genomes/hg19/hg19.json.
  - This behavior is evident from the log file in the data directory. (Clear the data/genomes directory, then launch IGV, and open the log file.)
- IGV will always create `%USERPROFILE%/.igv/java_arguments` if it does not already exist. This is not affected by the `--igvDirectory` argument.
- The default Genome server URL is https://s3.amazonaws.com/igv.org.genomes/genomes.tsv. See View > Preferences > Advanced.
- The default Data registry server is https://data.broadinstitute.org/igvdata/$$_dataServerRegistry.txt. See View > Preferences > Advanced. This specifies what available datasets are shown when you click File > Load from Server...
  - Example: https://data.broadinstitute.org/igvdata/hg38_dataServerRegistry.txt contains a link to an XML file (https://s3.amazonaws.com/igv.org.genomes/hg38/hg38_annotations.xml) that contains an organized set of annotations files for the hg38 genome, largely from the `igv.org.genomes` Amazon S3 bucket.
  - <span style="color: red">Most of these files *appear* to correspond to UCSC files, but their provenance is not explicitly clear.</span>
    - Example of minor discrepancy: RepeatMasker annotations for mouse mm10 genome
      - UCSC: either from [bigZips directory](https://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/latest/mm10.chromOut.tar.gz) or download via [UCSC Table Browser](http://genome.ucsc.edu/cgi-bin/hgTables) (group = "Variation and Repeats", track = "RepeatMasker", region = genome, filter = genoName does match "chr1") --> 72067 LINE loci on chromosome 1
      - IGV: from [Amazon S3 bucket](https://s3.amazonaws.com/igv.org.genomes/mm10/rmsk/mm10_rmsk_LINE.bed.gz) --> 72070 LINE loci on chromosome 1. The 3 extra LINE loci are all marked with a ? (as in "LINE?").

IGV/Broad Institute Amazon S3 buckets. A GET request directed at the bucket URL yields an XML file that gives a partial list of files available under that Amazon S3 bucket.
- https://s3.amazonaws.com/igv.broadinstitute.org/
  - The files listed in the XML file appear to all be annotation files
    (e.g., such as cytoband files or GENCODE GTF files).
  - Some genome specification files (\*.genome) appear to be stored in this Amazon S3 bucket, but are not listed in the XML file. See the default IGV Genome database (https://s3.amazonaws.com/igv.org.genomes/genomes.tsv) for examples.
- https://s3.amazonaws.com/igv.org.genomes/
  - The files listed in the XML file appear to be a mix of genome specification files (\*.genome or \*.json) and annotation files.
  - This appears to be more up-to-date:
    - Many genomes are provided as \*.json specifications. See https://github.com/igvteam/igv/wiki/Downloading-Hosted-Genomes-for-Offline-Use
    - The T2T CHM13 genome (chm13v2.0) is availabe here.
    - See also the Data registry server discussed above.

How tracks are loaded
- Indexing annotation files (e.g., via `tabix` from htslib) allow quick, memory-efficient loading, but indexed annotation files are not searchable [(see GitHub Issue)](https://github.com/igvteam/igv/issues/244).
- Non-indexed annotation files are loaded completely into memory.
